python/nnvm/frontend/onnx.py

```diff
@@ -4,87 +4,113 @@
 import tvm
 from .. import symbol as _sym
 from .. import graph as _graph
-from .. compiler import graph_util
+from ..compiler import graph_util
 from .common import get_nnvm_op, Renamer, AttrConverter as AttrCvt
 
 __all__ = ['from_onnx']
 
-def _revert_caffe2_pad(attr):
-    """Caffe2 require two times the normal padding."""
-    if len(attr) == 4:
-        attr = attr[:2]
-    elif len(attr) == 2:
-        pass
-    else:
-        raise ValueError("Invalid caffe2 type padding: {}".format(attr))
-    return attr
-
-def _math_name_picker(surfix):
-    def _impl(attr):
-        if attr.get('broadcast', 0):
-            return 'broadcast_' + surfix
-        return 'elemwise_' + surfix
-    return _impl
 
-def _broadcast_constraint():
-    def _broadcast_check(attrs):
-        if attrs.get('axis', None):
-            return False
-        return True
-    return _broadcast_check, "Specifying broadcast axis not allowed."
+class OnnxOpConverter(object):
+    """ A helper class for holding onnx op converters.
+    """
 
-def _dimension_picker(prefix, surfix=''):
-    def _impl(attr):
-        kernel = attr['kernel_shape']
-        if len(kernel) == 2:
-            return prefix + '2d' + surfix
+    @classmethod
+    def get_converter(cls, opset):
+        versions = [
+            int(d.replace('_impl_v', '')) for d in dir(cls) if '_impl_v' in d
+        ]
+        versions = sorted(versions + [opset])
+        version = versions[
+            max([i for i, v in enumerate(versions) if v == opset]) - 1]
+        if hasattr(cls, '_impl_v{}'.format(version)):
+            return getattr(cls, '_impl_v{}'.format(version))
         else:
-            raise NotImplementedError("Only 2d kernel supported.")
-    return _impl
+            raise NotImplementedError(
```

Comment:
Does it require rewriting _impl_vx for every onnx update?
How about a fallback support?

Reply:
@zhreshold 
If onnx updates special operator specification, we just need add one `_impl_` under that operator.
For example, if onnx add opset=7's Reshape op, we should add `_impl_v7` in class Reshape. Just that.